### DISCUSSION TITLE
"Invite" functionality

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#1206] [Q&A]
I'm trying to work out the best way to mimic invite functionality (using v0.9.0). i.e - i'm the owner of a workspace, and i'm going to invite a colleague to collaborate with me. Typically that involves creating a user with the bare minimum data, emailing a tokenised link to the user, then asking the user to confim and set a password.

This is what i'm doing at the moment:

1. Create the user: `client.collection('users').create(...)` <- This is problematic, because i _have_ to pass a password. But for now i'm setting a randomly generated one.
2. Request verification: I'm using this as an alternative to an invite-type method given it doesn't exist:  `client.collection('users').requestVerification(_newlyCreatedEmailFromStep1_)`.
3. Once email is received and user opens link, i'm using the token to confirm verification: `client.collection("users").confirmVerification(route.query.token)`
4. Finally, i'm trying to reset the password: `client.collection("users").confirmPasswordReset(...)`

Step 4 errors because the token used has already been used by step 3 - so i'd have to potentially log the user behind the scenes after step 3, grab the token, and re-run step 4. 

Just not sure if this whole things is feasible with Pocketbase 🤷‍♂️

Any thoughts or ideas?

**Comment by ganigeorgiev** [comment]
The secrets for the auth, verification and password-reset tokens are different so you can't use a verification token in a confirm password-reset request.

The easiest workaround would be to send a password reset email to the user with a custom invitation template.
Another approach could be to send a custom email with link to your registration page (you can pass default user data as query params and load them client-side after the user click on the email link).

But all of the above, requires to some extend to use [PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and [the available event hooks](https://pocketbase.io/docs/event-hooks/).
In v0.10 or v0.11 there will be support to extend and bind to the PocketBase app hooks with JS.


Eventually, we could also try to implement a generic invitation flow, but I'm not sure how to structure and define the rules who'll able to send the invitation emails (the easiest solution would be to allow only admins). Suggestions and example(s) with the desired workflow are welcome.

**Reply by charbs-io** [reply]
Thanks @ganigeorgiev . I'll play around with a few solutions, including exploring extending the framework option - something i haven't looked into before. 

A potential flow I can foresee would be to include two new API methods on the `users` collection: 

```
await pb.collection('users').sendInvitation('test@example.com');
```

This would: 
1. Send invite to user (with its own email template given the uniqueness of the email). The email would include a unique token for the intended email.
2. Create a brand new user in the `users` collection in the background with `verified` set to false (the `password` and `passwordConfirm` are currently required when `create` user is invoked, but perhaps something can be done in this case to make them optional - thus only needing the `email`).

The invited user would be expected to access the tokenised link and:
1. Set a password
2. Set a confirm password
3. (optional) set other things like name, avatar etc (these would depend on the application and can be extended how a developer sees fit)
4. Submit a form of sorts which would invoke the second method...


```
await pb.collection('users').acceptInvitation('test@example.com', {token}, {password}, {passwordConfirm});
```

This second method would:
1. Set the user record `verified` value to `true` - i.e the user is now verified
2. Set the password for the user
3. Return a similar response to that of the `authWithPassword` method - i.e {token: ..., record:...} - that means the app developer can log the user in as soon as they've accepted the invitation.


I realise there's a lot more stuff to do with higher priority, but a lot of SaaS requires user flows very similar to the above due to the nature of collaborative apps being popular.

Thanks!

Edit: Perhaps the `verified` flag is best left alone, and maybe another flag is used? Not sure if that's the best design as you don't want a flag that is super specific to that flow 🤷‍♂️

**Reply by ganigeorgiev** [reply]
Thank you for the details. I'll have to think a little more on this.

As mentioned previously, my main concern is how to structure and define who'll be able to send the invitation, because we can't just allow the invite action to be public and create users on the fly since it could get abused and eventually conflict with the collection Create API rule.

**Reply by charbs-io** [reply]
Yes that’s fair enough. I forgot to include that above. The way I was planning on doing this was by implementing an API rule (through the existing admin UI) on the users collection’s create method to ensure only 
1. Logged in users can create users (existing functionality and possible)
2. Logged in users who are part of the workspace with a role of `owner` (`workspace` is a collection I created that happens to have a relation with users and another collection called userroles) 

but I understand #2 above is again too specific to be generic. 

Thanks

**Reply by aonghas** [reply]
Hey I was wondering where you ended up getting with this, or what the latest ideas are about implementing a similar flow with the latest version of Pocketbase?

**Reply by ganigeorgiev** [reply]
@aonghas This is a very old discussion and a lot of things has changed since then but there is no builtin "invite" functionality yet (_it is buried somewhere in my never ending todo for later considerations bur remains a very low priority_). If you have other questions please consider opening a new discussion with more details about your use case.
